core/core.mk

```diff
@@ -60,6 +60,10 @@ ifneq ($(CFG_TEE_CORE_DEBUG),y)
 cppflags$(sm)  += -DNDEBUG
 endif
 
+ifeq ($(CFG_WARN_DECL_AFTER_STATEMENT),y)
+cflags$(sm)	+= $(call cc-option,-Wdeclaration-after-statement)
+endif
```

Comment:
Since you have fixed all occurrences in the prior patches, perhaps this could be done always (no need for a CFG_)?

Reply:
I was thinking the same at first, but there's the problem with out of tree code. The optee_os/core part may not be so bad but this would have the potential to affect TAs built with TA dev-kit. So just as with `CFG_WERROR` I thought would be safer to let this be disabled by default.

Reply:
This sets the flags for the core only, so there should be no problem for external TAs. And IMO in-tree TAs should have the flag too:
```diff
diff --git a/ta/mk/build-user-ta.mk b/ta/mk/build-user-ta.mk
index 8138d474..7a8c1c1c 100644
--- a/ta/mk/build-user-ta.mk
+++ b/ta/mk/build-user-ta.mk
@@ -30,6 +30,7 @@ include mk/$(COMPILER_$(sm)).mk
 
 cppflags$(sm)  := $(cppflags$(ta-target)) -I$(ta-dev-kit-dir$(sm))/include
 cflags$(sm)    := $(cflags$(ta-target))
+cflags$(sm)    += $(call cc-option,-Wdeclaration-after-statement)
 aflags$(sm)    := $(aflags$(ta-target))
 
 ifeq ($(CFG_ULIBS_SHARED),y)
```

Another option would be to add it to `mk/compile.mk` only so that any module has it, but guarded with the `CFG_` so that it can be disabled when TA dev kit is used externally:
```diff
diff --git a/mk/compile.mk b/mk/compile.mk
index 9803f7e2..1e60cfe6 100644
--- a/mk/compile.mk
+++ b/mk/compile.mk
@@ -36,6 +36,9 @@ comp-cflags-warns-high = \
        -Wshadow -Wstrict-prototypes -Wswitch-default \
        -Wwrite-strings \
        -Wno-missing-field-initializers -Wno-format-zero-length
+ifeq ($(CFG_WARN_DECL_AFTER_STATEMENT),y)
+comp-cflags-warn-high += $(call cc-option,-Wdeclaration-after-statement)
+endif
 comp-cflags-warns-medium = \
        -Wredundant-decls
 comp-cflags-warns-low = \
```

Reply:
I like the second option better, but we could enable `CFG_WARN_DECL_AFTER_STATEMENT` by default instead. I'll give a try, stay tuned... :-)